tools/k8s_versions/update_k8s_versions.go

```diff
@@ -0,0 +1,285 @@
+
+	logDebug("Found minikube versions: %s", minikuveVersions)
+
+	var latestMinikubeVersions []string
+	// the minikube version slice is sorted, break when first cycle match is found
+	for _, k8sVersion := range k8sVersions {
+		for _, minikubeVersion := range minikuveVersions {
+			if strings.Contains(minikubeVersion, k8sVersion.Cycle) {
+				latestMinikubeVersions = append(latestMinikubeVersions, strings.TrimSpace(minikubeVersion))
+				break
+			}
+		}
+	}
+
+	return latestMinikubeVersions, nil
+}
+
+// getLatestSupportedKindImages iterates through the K8s supported versions and find the latest kind
+// tag that supports that version
+func getLatestSupportedKindImages(url string, k8sVersions []KubernetesVersion) ([]string, error) {
+	var supportedKindVersions []string
+	for _, k8sVersion := range k8sVersions {
+		tag := k8sVersion.Latest
+		for {
+			exists, err := imageTagExists(url, tag)
+			if err != nil {
+				return supportedKindVersions, fmt.Errorf("failed to check image tag existence: %w", err)
+			}
+			if exists {
+				supportedKindVersions = append(supportedKindVersions, "v"+tag)
+				break
+			}
+			tag, err = decrementMinorMinorVersion(tag)
+			if err != nil {
+				// It's possible that kind still does not have a tag for new versions, break the loop and
+				// process other k8s versions
+				if strings.Contains(err.Error(), "minor version cannot be decremented below 0") {
+					logDebug("No kind image found for k8s version %s", k8sVersion.Cycle)
+					break
+				}
+				return supportedKindVersions, fmt.Errorf("failed to decrement k8sVersion: %w", err)
+			}
+		}
+	}
+	return supportedKindVersions, nil
+}
+
+func imageTagExists(url string, tag string) (bool, error) {
+	body, err := getRequestBody(url + tag)
+	if err != nil {
+		return false, fmt.Errorf("failed to get image tag: %w", err)
+	}
+
+	var kindImage DockerImage
+	if err := json.Unmarshal(body, &kindImage); err != nil {
+		return false, fmt.Errorf("failed to unmarshal JSON: %w", err)
+	}
+
+	if kindImage.Count > 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+func decrementMinorMinorVersion(version string) (string, error) {
+	parts := strings.Split(version, ".")
+	if len(parts) < 3 {
+		return "", fmt.Errorf("version does not have a minor version: %s", version)
+	}
+
+	minor, err := strconv.Atoi(parts[2])
+	if err != nil {
+		return "", fmt.Errorf("invalid minor version: %s", parts[1])
+	}
+
+	if minor == 0 {
+		return "", fmt.Errorf("minor version cannot be decremented below 0")
+	}
+
+	parts[2] = strconv.Itoa(minor - 1)
+	return strings.Join(parts, "."), nil
+}
+
+func updateMatrixFile(filePath string, kindVersions []string, minikubeVersions []string) error {
+	content, err := os.ReadFile(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to read file: %w", err)
+	}
+
+	var testMatrix map[string]map[string][]string
+	if err := json.Unmarshal(content, &testMatrix); err != nil {
```

Comment:
Same as other comment `err` ends up getting overshadowed here, we should use `=` instead through the rest of the method.